File: app/src/main/java/org/adaway/ui/home/HomeViewModel.java
package org.adaway.ui.home;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MediatorLiveData;
import androidx.lifecycle.MutableLiveData;

import org.adaway.AdAwayApplication;
import org.adaway.db.AppDatabase;
import org.adaway.db.dao.HostListItemDao;
import org.adaway.db.dao.HostsSourceDao;
import org.adaway.model.adblocking.AdBlockModel;
import org.adaway.model.error.HostError;
import org.adaway.model.error.HostErrorException;
import org.adaway.model.source.SourceModel;
import org.adaway.model.update.Manifest;
import org.adaway.model.update.UpdateModel;
import org.adaway.util.AppExecutors;

import timber.log.Timber;

/**
 * This class is an {@link AndroidViewModel} for the {@link HomeActivity} cards.
 *
 * @author Bruce BUJON (bruce.bujon(at)gmail(dot)com)
 */
public class HomeViewModel extends AndroidViewModel {
    private static final String TAG = "NextViewModel";
    private static final AppExecutors EXECUTORS = AppExecutors.getInstance();

    private final SourceModel sourceModel;
    private final AdBlockModel adBlockModel;
    private final UpdateModel updateModel;

    private final HostsSourceDao hostsSourceDao;
    private final HostListItemDao hostListItemDao;

    private final MutableLiveData<Boolean> pending;
    private final MediatorLiveData<String> state;
    private final MutableLiveData<HostError> error;

    public HomeViewModel(@NonNull Application application) {
        super(application);
        AdAwayApplication awayApplication = (AdAwayApplication) application;
        this.sourceModel = awayApplication.getSourceModel();
        this.adBlockModel = awayApplication.getAdBlockModel();
        this.updateModel = awayApplication.getUpdateModel();

        AppDatabase database = AppDatabase.getInstance(application);
        this.hostsSourceDao = database.hostsSourceDao();
        this.hostListItemDao = database.hostsListItemDao();

        this.pending = new MutableLiveData<>(false);
        this.state = new MediatorLiveData<>();
        this.state.addSource(this.sourceModel.getState(), this.state::setValue);
        this.state.addSource(this.adBlockModel.getState(), this.state::setValue);
        this.error = new MutableLiveData<>();
    }

    private static boolean isTrue(LiveData<Boolean> liveData) {
        Boolean value = liveData.getValue();
        return value != null && value;
    }

    public LiveData<Boolean> isAdBlocked() {
        return this.adBlockModel.isApplied();
    }

    public LiveData<Boolean> isUpdateAvailable() {
        return this.sourceModel.isUpdateAvailable();
    }

    public String getVersionName() {
        return this.updateModel.getVersionName();
    }

    public LiveData<Manifest> getAppManifest() {
        return this.updateModel.getManifest();
    }

    public LiveData<Integer> getBlockedHostCount() {
        return this.hostListItemDao.getBlockedHostCount();
    }

    public LiveData<Integer> getAllowedHostCount() {
        return this.hostListItemDao.getAllowedHostCount();
    }

    public LiveData<Integer> getRedirectHostCount() {
        return this.hostListItemDao.getRedirectHostCount();
    }

    public LiveData<Integer> getUpToDateSourceCount() {
        return this.hostsSourceDao.countUpToDate();
    }

    public LiveData<Integer> getOutdatedSourceCount() {
        return this.hostsSourceDao.countOutdated();
    }

    public LiveData<Boolean> getPending() {
        return this.pending;
    }

    public LiveData<String> getState() {
        return this.state;
    }

    public LiveData<HostError> getError() {
        return this.error;
    }

    public void checkForAppUpdate() {
        EXECUTORS.networkIO().execute(this.updateModel::checkForUpdate);
    }

    public void toggleAdBlocking() {
        if (isTrue(this.pending)) {
            return;
        }
        EXECUTORS.diskIO().execute(() -> {
            try {
                this.pending.postValue(true);
                if (isTrue(this.adBlockModel.isApplied())) {
                    this.adBlockModel.revert();
                } else {
                    this.adBlockModel.apply();
                }
            } catch (HostErrorException exception) {
                Timber.w(exception, "Failed to toggle ad blocking.");
                this.error.postValue(exception.getError());
            } finally {
                this.pending.postValue(false);
            }
        });
    }

    public void update() {
        if (isTrue(this.pending)) {
            return;
        }
        EXECUTORS.networkIO().execute(() -> {
            try {
                this.pending.postValue(true);
                this.sourceModel.checkForUpdate();
            } catch (HostErrorException exception) {
                Timber.w(exception, "Failed to update.");
                this.error.postValue(exception.getError());
            } finally {
                this.pending.postValue(false);
            }
        });
    }

    public void sync() {
        if (isTrue(this.pending)) {
            return;
        }
        EXECUTORS.networkIO().execute(() -> {
            try {
                this.pending.postValue(true);
                this.sourceModel.retrieveHostsSources();
                this.adBlockModel.apply();
            } catch (HostErrorException exception) {
                Timber.w(exception, "Failed to sync.");
                this.error.postValue(exception.getError());
            } finally {
                this.pending.postValue(false);
            }
        });
    }

    public void enableAllSources() {
        EXECUTORS.diskIO().execute(() -> {
            if (this.sourceModel.enableAllSources()) {
                sync();
            }
        });
    }
}

File: app/src/main/java/org/adaway/ui/home/HomeViewModel.kt
package org.adaway.ui.home

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.asFlow
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.merge
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.adaway.AdAwayApplication
import org.adaway.db.AppDatabase
import org.adaway.db.dao.HostListItemDao
import org.adaway.db.dao.HostsSourceDao
import org.adaway.helper.NotificationHelper
import org.adaway.helper.PreferenceHelper
import org.adaway.model.adblocking.AdBlockMethod
import org.adaway.model.adblocking.AdBlockModel
import org.adaway.model.error.HostError
import org.adaway.model.error.HostErrorException
import org.adaway.model.source.SourceModel
import org.adaway.model.update.Manifest
import org.adaway.model.update.UpdateModel
import org.adaway.vpn.VpnStatusRepository
import timber.log.Timber

/**
 * This class is an [AndroidViewModel] for the [HomeActivity] cards.
 *
 * @author Bruce BUJON (bruce.bujon(at)gmail(dot)com)
 */
class HomeViewModel(application: Application) : AndroidViewModel(application) {

    private val sourceModel: SourceModel
    private val adBlockModel: AdBlockModel
    private val updateModel: UpdateModel

    private val hostsSourceDao: HostsSourceDao
    private val hostListItemDao: HostListItemDao

    private val _pending = MutableStateFlow(false)
    val pending: StateFlow<Boolean> = _pending

    private val _error = MutableSharedFlow<HostError>()
    val error: SharedFlow<HostError> = _error

    init {
        val awayApplication = application as AdAwayApplication
        sourceModel = awayApplication.sourceModel
        adBlockModel = awayApplication.adBlockModel
        updateModel = awayApplication.updateModel

        val database = AppDatabase.getInstance(application)
        hostsSourceDao = database.hostsSourceDao()
        hostListItemDao = database.hostsListItemDao()

        VpnStatusRepository.update(PreferenceHelper.getVpnServiceStatus(application))
    }

    val state: StateFlow<String> = merge(
        sourceModel.state.asFlow(),
        adBlockModel.state.asFlow()
    ).stateIn(viewModelScope, SharingStarted.WhileSubscribed(FLOW_STOP_TIMEOUT_MILLIS), "")

    val adBlocked: StateFlow<Boolean> = combine(
        adBlockModel.isApplied.asFlow().map { it == true },
        VpnStatusRepository.status
    ) { applied, vpnStatus ->
        if (adBlockModel.method == AdBlockMethod.VPN) {
            vpnStatus.isStarted
        } else {
            applied
        }
    }.stateIn(
        viewModelScope,
        SharingStarted.WhileSubscribed(FLOW_STOP_TIMEOUT_MILLIS),
        adBlockModel.isApplied.value == true
    )

    val updateAvailable: StateFlow<Boolean> = sourceModel.isUpdateAvailable
        .asFlow()
        .map { it == true }
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(FLOW_STOP_TIMEOUT_MILLIS), false)

    val versionName: String get() = updateModel.versionName

    val appManifest: StateFlow<Manifest?> = updateModel.manifest
        .asFlow()
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(FLOW_STOP_TIMEOUT_MILLIS), updateModel.manifest.value)

    val blockedHostCount: StateFlow<Int> = hostListItemDao.getBlockedHostCount()
        .asFlow()
        .map { it }
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(FLOW_STOP_TIMEOUT_MILLIS), 0)

    val allowedHostCount: StateFlow<Int> = hostListItemDao.getAllowedHostCount()
        .asFlow()
        .map { it }
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(FLOW_STOP_TIMEOUT_MILLIS), 0)

    val redirectHostCount: StateFlow<Int> = hostListItemDao.getRedirectHostCount()
        .asFlow()
        .map { it }
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(FLOW_STOP_TIMEOUT_MILLIS), 0)

    val upToDateSourceCount: StateFlow<Int> = hostsSourceDao.countUpToDate()
        .asFlow()
        .map { it }
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(FLOW_STOP_TIMEOUT_MILLIS), 0)

    val outdatedSourceCount: StateFlow<Int> = hostsSourceDao.countOutdated()
        .asFlow()
        .map { it }
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(FLOW_STOP_TIMEOUT_MILLIS), 0)

    fun checkForAppUpdate() {
        viewModelScope.launch(Dispatchers.IO) {
            updateModel.checkForUpdate()
        }
    }

    fun toggleAdBlocking() {
        if (_pending.value) {
            return
        }
        viewModelScope.launch {
            try {
                _pending.value = true
                withContext(Dispatchers.IO) {
                    if (adBlocked.value) {
                        adBlockModel.revert()
                    } else {
                        adBlockModel.apply()
                    }
                }
            } catch (exception: HostErrorException) {
                Timber.w(exception, "Failed to toggle ad blocking.")
                _error.emit(exception.error)
            } finally {
                _pending.value = false
            }
        }
    }

    fun update() {
        if (_pending.value) {
            return
        }
        viewModelScope.launch {
            try {
                _pending.value = true
                withContext(Dispatchers.IO) {
                    sourceModel.checkForUpdate()
                }
            } catch (exception: HostErrorException) {
                Timber.w(exception, "Failed to update.")
                _error.emit(exception.error)
            } finally {
                _pending.value = false
            }
        }
    }

    fun sync() {
        if (_pending.value) {
            return
        }
        viewModelScope.launch {
            val application = getApplication<Application>()
            try {
                _pending.value = true
                NotificationHelper.showUpdateHostsProgressNotification(application)
                withContext(Dispatchers.IO) {
                    sourceModel.retrieveHostsSources()
                    adBlockModel.apply()
                }
            } catch (exception: HostErrorException) {
                Timber.w(exception, "Failed to sync.")
                _error.emit(exception.error)
            } finally {
                NotificationHelper.clearUpdateHostsProgressNotification(application)
                _pending.value = false
            }
        }
    }

    fun enableAllSources() {
        viewModelScope.launch {
            val enabled = withContext(Dispatchers.IO) {
                sourceModel.enableAllSources()
            }
            if (enabled) {
                sync()
            }
        }
    }

    companion object {
        private const val FLOW_STOP_TIMEOUT_MILLIS = 5_000L
    }
}
